File: 11_动态规划/力扣加加/数位dp/tmp/902. 最大为 N 的数字组合.ts
/**
 * @param {string[]} digits
 * @param {number} n 1 <= N <= 10^9
 * @return {number}
 * 我们有一组排序的数字 D，它是  {'1','2','3','4','5','6','7','8','9'} 的非空子集。
 * 返回可以用 D 中的数字写出的小于或等于 N 的正整数的数目。
 * @description 必须用数学方法 否则超时
 */
const atMostNGivenDigitSet = function (digits: string[], n: number): number {
  const nums = digits.map(Number)
  const len = n.toString().length
  let res = 0

  // 小于len位的
  for (let i = 1; i < len; i++) {
    res += nums.length ** i
  }

  // 等于len位时 n的高位向低位对比
  for (let i = 0; i < len; i++) {
    const upper = Number(n.toString()[i])
    // 在此位能取的数
    const lessThan = nums.filter(v => v < upper).length
    res += lessThan * digits.length ** (len - i - 1)
    console.log(res, lessThan)

    // 如果存在upper 则不必继续看了比n小的已经看完
    if (!nums.includes(upper)) break
    if (i === len - 1) res++ // 等于n
  }

  return res
}

console.log(atMostNGivenDigitSet(['1', '3', '5', '7'], 786))
console.log(atMostNGivenDigitSet(['1', '3', '5', '7'], 100))

export {}

File: 11_动态规划/dp分类/数位dp/902. 最大为 N 的数字组合.py
from functools import lru_cache
from typing import List


def cal(upper: int, digits: List[int]) -> int:
    @lru_cache(None)
    def dfs(pos: int, hasLeadingZero: int, isLimit: bool) -> int:
        """当前在第pos位，hasLeadingZero表示有前导0，isLimit表示是否贴合上界"""
        if pos == len(nums):
            return not hasLeadingZero

        res = 0
        up = nums[pos] if isLimit else 9
        for cur in range(up + 1):
            if hasLeadingZero and cur == 0:
                res += dfs(pos + 1, True, (isLimit and cur == up))
            else:
                if cur not in digits:
                    continue
                res += dfs(pos + 1, False, (isLimit and cur == up))
        return res

    nums = list(map(int, str(upper)))
    return dfs(0, True, True)


class Solution:
    def atMostNGivenDigitSet(self, digits: List[str], n: int) -> int:
        """返回 可以生成的小于或等于给定整数 n 的正整数的个数 。"""
        return cal(n, list(map(int, digits)))


print(Solution().atMostNGivenDigitSet(digits=["1", "3", "5", "7"], n=100))
